packages/grafana-ui/src/components/Table/TableInputCSV.tsx

```diff
@@ -0,0 +1,170 @@
+import React from 'react';
+import debounce from 'lodash/debounce';
+import Papa, { ParseError, ParseMeta } from 'papaparse';
+import { TableData, Column } from '../../types/data';
+
+// Subset of all parse options
+export interface ParseConfig {
+  delimiter?: string; // default: ","
+  newline?: string; // default: "\r\n"
+  quoteChar?: string; // default: '"'
+  encoding?: string; // default: ""
+  comments?: boolean | string; // default: false
+}
+
+export interface ParseResults {
+  table: TableData;
+  meta: ParseMeta;
+  errors: ParseError[];
+}
+
+// This mutates the table structure!
+export function checkAndFix(table: TableData): number {
+  let cols = table.columns.length;
+  let different = 0;
+  table.rows.forEach(row => {
+    if (cols !== row.length) {
+      different++;
+      cols = Math.max(cols, row.length);
+    }
+  });
+  if (different > 0) {
+    if (cols !== table.columns.length) {
+      const diff = cols - table.columns.length;
+      for (let i = 0; i < diff; i++) {
+        table.columns.push({
+          text: 'Column ' + table.columns.length,
+        });
+      }
+    }
+    table.rows.forEach(row => {
+      const diff = cols - row.length;
+      for (let i = 0; i < diff; i++) {
+        row.push(null);
+      }
+    });
+  }
+  return different;
+}
+
+export function parseCSV(text: string, config?: ParseConfig): ParseResults {
+  const results = Papa.parse(text, { ...config, dynamicTyping: true, skipEmptyLines: true });
+
+  const { data, meta, errors } = results;
+  if (!data || data.length < 1) {
+    if (!text) {
+      errors.length = 0; // clear other errors
+    }
+    errors.push({
+      type: 'warning', // A generalization of the error
+      message: 'Empty Data',
+      code: 'empty',
+      row: 0,
+    });
+    return {
+      table: {
+        columns: [],
+        rows: [],
+        type: 'table',
+        columnMap: {},
+      } as TableData,
+      meta,
+      errors,
+    };
+  }
+
+  const first = results.data.shift();
```

Comment:
`first` what? ;) I assume it'a a `tableHeader`.

Reply:
:)